hbase_storage/src/main/java/com/splicemachine/timestamp/hbase/ZkTimestampBlockManager.java

```diff
@@ -43,11 +43,17 @@ public ZkTimestampBlockManager(RecoverableZooKeeper rzk,String blockNode) {
         this.blockNode = blockNode;
     }
 
-
-    public void reserveNextBlock(long nextMax) throws TimestampIOException{
-        byte[] data = Bytes.toBytes(nextMax);
+    @Override
+    public long reserveNextBlock(long nextMax) throws TimestampIOException{
         try {
-            rzk.setData(blockNode, data, -1 /* version */); // durably reserve the next block
+            byte[] bs = rzk.getData(blockNode, false, null);
```

Comment:
Could we use the version parameter in .setData() to make sure we only update it if it matches the expected version?

I don't think we should be playing catch up here, if somebody else updated the stored timestamp we should bail out, either somebody else is the new TimestampOracle or we are in the middle of a restore were we shouldn't be creating new transactions.

Reply:
It is possible that a timestamp is increased on a replica. We may want to catch up in this case.

Reply:
I thought replicas were read only and we were not increasing timestamps there, how would that happen?

Reply:
The replica applies WALs and compare the progress with the snapshot of master cluster. If it has replicated all WALs for a snapshot, the replica bumps up timestamp to the snapshot's timestamp.

Reply:
When we got the timezone correct (matching the timezone on the backup) on the cluster where we were testing this, we did not need to apply the workaround.  Seems to be timezone related.  So, to Oleg's point, is this PR addressing a symptom rather than the problem (backup restoring to a different timezone)?

Reply:
I see that replication reads/updates the same Zookeeper node (_/transactions/maxReservedTimestamp_)  directly, bypassing TimestampOracle. That violates many assumptions made in TimestampOracle code.
TimestampOracle itself is not free from subtle data races. _timestampCounter.set() can not be reliably used in multithreaded environment (it's atomic all right, but can't maintain integrity) and should be replaced with compare-and-swap.
I believe replication approach should be revised.

Reply:
I don't think it's TZ related, I think it's mostly dependent on how long the cluster is running after the restore finishes, and there's a bit of (bad) luck involved. Some attempts the other day with the incorrect TZ didn't hit this problem either, I think it's just coincidence that we didn't hit it now.